gui/biomes.lua

```diff
@@ -0,0 +1,379 @@
+    for i = #choices + 1, #biomeList do
+        local biomeExt = biomeList[i]
+        table.insert(choices, {
+            text = ([[%s: %s]]):format(biomeExt.char, GetBiomeName(biomeExt.biome, biomeExt.typeId)),
+            icon = self:get_icon_pen_callback(#choices+1),
+            biomeTypeId = biomeExt.typeId,
+            biome = biomeExt.biome,
+        })
+    end
+    self.list:setChoices(choices)
+end
+
+do -- implementation of onMouseHoverEntry(idx, option)
+    function BiomeVisualizerLegend:onRenderFrame(dc, rect)
+        BiomeVisualizerLegend.super.onRenderFrame(self, dc, rect)
+
+        local list = self.list
+        local currentHoverIx = list:getIdxUnderMouse()
+        local oldIx = self.HoverIndex
+        if currentHoverIx ~= oldIx then
+            self.HoverIndex = currentHoverIx
+            if self.onMouseHoverEntry then
+                local choices = list:getChoices()
+                self:onMouseHoverEntry(currentHoverIx, choices[currentHoverIx])
+            end
+        end
+    end
+end
+
+local function add_field_text(lines, biome, field_name)
+    lines[#lines+1] = ("%s: %s"):format(field_name, biome[field_name])
+    lines[#lines+1] = NEWLINE
+end
+
+local function get_tooltip_text(option)
+    local text = {}
+    text[#text+1] = ("type: %s"):format(df.biome_type[option.biomeTypeId])
+    text[#text+1] = NEWLINE
+
+    local biome = option.biome
+
+    add_field_text(text, biome, "savagery")
+    add_field_text(text, biome, "evilness")
+    table.insert(text, NEWLINE)
+
+    add_field_text(text, biome, "elevation")
+    add_field_text(text, biome, "rainfall")
+    add_field_text(text, biome, "drainage")
+    add_field_text(text, biome, "vegetation")
+    add_field_text(text, biome, "temperature")
+    add_field_text(text, biome, "volcanism")
+    table.insert(text, NEWLINE)
+
+    local flags = biome.flags
+    if flags.is_lake then
+        text[#text+1] = "lake"
+        text[#text+1] = NEWLINE
+    end
+    if flags.is_brook then
+        text[#text+1] = "brook"
+        text[#text+1] = NEWLINE
+    end
+
+    return text
+end
+
+function BiomeVisualizerLegend:onMouseHoverEntry(idx, option)
+    self:ShowTooltip(option or self.SelectedOption)
+end
+
+function BiomeVisualizerLegend:ShowTooltip(option)
+    local text = get_tooltip_text(option)
+
+    local tooltip_panel = self.tooltip_panel
+    local lbl = tooltip_panel.subviews.label
+
+    lbl:setText(text)
+end
+
+--------------------------------------------------------------------------------
+
+if RELOAD then BiomeVisualizer = nil end
+BiomeVisualizer = defclass(BiomeVisualizer, gui.ZScreen)
+BiomeVisualizer.ATTRS{
+    focus_path='BiomeVisualizer',
+    pass_movement_keys=true,
+}
+
+function BiomeVisualizer:init()
+    local legend = BiomeVisualizerLegend{view_id = 'legend'}
+    self:addviews{legend}
+end
+
+function BiomeVisualizer:onRenderFrame(dc, rect)
+    BiomeVisualizer.super.onRenderFrame(self, dc, rect)
+
```

Comment:
```suggestion

    if not dfhack.screen.inGraphicsMode() and not gui.blink_visible(500) then
        return
    end

```

Reply:
so the player can still see the underlying map in ascii mode. otherwise it looks like this:

![image](https://github.com/DFHack/scripts/assets/977482/86bf8bd7-acdd-4d18-82b7-e6b6d269be02)

Reply:
Ah, so *that*'s how the blinking is done :) Good to know